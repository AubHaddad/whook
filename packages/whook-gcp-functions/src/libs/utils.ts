import { YError } from 'yerror';
import path from 'path';
import type { LogService } from 'common-services';

export async function loadFunction(
  {
    PROJECT_DIR,
    log,
  }: {
    PROJECT_DIR: string;
    log: LogService;
  },
  target: string,
  operationId: string,
  type: string,
  extension = '.mjs',
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): Promise<any> {
  const modulePath = path.join(
    PROJECT_DIR,
    'builds',
    target,
    operationId,
    type + extension,
  );

  log('debug', `⛏️ - Loading function module at path "${modulePath}".`);

  try {
    const module = await import(modulePath);

    if (!module) {
      throw new YError('E_MODULE_NOT_FOUND', module);
    }

    if (!module.default) {
      throw new YError('E_LAMBDA_NOT_FOUND', module, Object.keys(module));
    }

    return module.default;
  } catch (err) {
    throw YError.wrap(err as Error, 'E_LAMBDA_LOAD');
  }
}
