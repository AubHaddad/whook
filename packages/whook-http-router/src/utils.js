import OpenAPIParser from 'swagger-parser';
import YError from 'yerror';

/**
 * Flatten the inputed OpenAPI file
 *  object
 * @param  {Object} API
 * An Object containing a parser OpenAPI JSON
 * @return {Object}
 * The flattened OpenAPI definition
 */
export async function flattenOpenAPI(API) {
  try {
    const parser = new OpenAPIParser();

    // Currently the OpenAPI parser changes the API in place
    //  this is why we're deep cloning it here
    const result = await parser.dereference(JSON.parse(JSON.stringify(API)));

    return result;
  } catch (err) {
    throw YError.wrap(err, 'E_BAD_OPEN_API');
  }
}

/**
 * Return a OpenAPI operation in a more
 *  convenient way to iterate onto its
 *  operations
 * @param  {Object} API
 * The flattened OpenAPI defition
 * @return {Array}
 * An array of all the OpenAPI operations
 * @example
 * getOpenAPIOperations(API)
 * .map((operation) => {
 *    const { path, method, operationId, parameters } = operation;
 *
 *   // Do something with that operation
 * });
 */
export function getOpenAPIOperations(API) {
  return Object.keys(API.paths).reduce(
    (operations, path) =>
      Object.keys(API.paths[path]).reduce(
        (operations, method) =>
          operations.concat({
            path,
            method,
            ...API.paths[path][method],
          }),
        operations,
      ),
    [],
  );
}
