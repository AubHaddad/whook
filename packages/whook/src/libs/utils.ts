// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function noop(..._args: unknown[]): void {
  return undefined;
}

export function identity<T>(id: T): T {
  return id;
}

export function compose<
  T extends (...args: unknown[]) => unknown = (...args: unknown[]) => unknown,
>(...fns: ((...args: unknown[]) => unknown)[]): T {
  return fns.reduce(
    (f, g) =>
      (...args: unknown[]) =>
        f(g(...args)),
  ) as T;
}

export function pipe(
  ...fns: ((...args: unknown[]) => unknown)[]
): (...args: unknown[]) => unknown {
  return compose.apply(compose, fns.reverse());
}
