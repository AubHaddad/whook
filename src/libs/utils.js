export function noop() {}

export function identity(id) {
  return id;
}

export function pipe(...fns) {
  return compose.apply(compose, fns.reverse());
}

export function compose(...fns) {
  return fns.reduce((f, g) => (...args) => f(g(...args)));
}
