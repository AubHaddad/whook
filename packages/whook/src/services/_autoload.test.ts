/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, it, beforeEach, jest, expect } from '@jest/globals';
import initAutoload from './_autoload.js';
import { service } from 'knifecycle';
import { YError } from 'yerror';
import { identity } from '../libs/utils.js';
import type { ImporterService, LogService } from 'common-services';
import type {
  ServiceInitializer,
  Dependencies,
  Service,
  Injector,
} from 'knifecycle';

describe('$autoload', () => {
  const log = jest.fn<LogService>();
  const $injector = jest.fn<Injector<any>>();
  const importer = jest.fn<ImporterService<any>>();
  const resolve = jest.fn();

  beforeEach(() => {
    log.mockReset();
    $injector.mockReset();
    importer.mockReset();
    resolve.mockReset();
  });

  describe('should work', () => {
    it('for configs', async () => {
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/services/CONFIGS.js',
      );
      importer.mockResolvedValueOnce({
        default: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('CONFIGS');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for a config constant', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('CONFIG');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for API', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/services/API.js',
      );
      importer.mockImplementationOnce(async () => ({
        default: service(async () => ({ info: {} }), 'API'),
      }));

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('API');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for SERVICE_NAME_MAP', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js',
      );
      importer.mockImplementationOnce(async () => ({
        default: service(async () => ({ info: {} }), 'SERVICE_NAME_MAP'),
      }));

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('SERVICE_NAME_MAP');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for handlers hash', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: {
          openapi: '3.0.2',
          info: {
            version: '1.0.0',
            title: 'Sample OpenAPI',
            description: 'A sample OpenAPI file for testing purpose.',
          },
          paths: {
            '/ping': {
              get: {
                operationId: 'getPing',
                summary: "Checks API's availability.",
                responses: {
                  '200': {
                    description: 'Pong',
                  },
                },
              },
            },
          },
        },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      importer.mockResolvedValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('HANDLERS');

      expect({
        result,
        HANDLERS: await (
          result.initializer as ServiceInitializer<Dependencies, Service>
        )({
          getPing: () => undefined,
        }),
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      importer.mockResolvedValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for name mapped handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPingMock.js',
      );
      importer.mockResolvedValueOnce({
        default: service(
          async () => async () => ({ status: 200 }),
          'getPingMock',
        ),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {
          getPing: 'getPingMock',
        },
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for name mapped handlers with dynamic SERVICE_NAME_MAP', async () => {
      $injector.mockResolvedValueOnce({
        SERVICE_NAME_MAP: {
          getPing: 'getPingMock',
        },
      });
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPingMock.js',
      );
      importer.mockResolvedValueOnce({
        default: service(
          async () => async () => ({ status: 200 }),
          'getPingMock',
        ),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for path mapped handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      importer.mockResolvedValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {
          getPing: '/home/whoami/my-other-project/src/handlers/getPing.js',
        },
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for handlers in sub plugins', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockImplementationOnce(() => {
        throw new YError('E_BAD_MODULE');
      });
      resolve.mockImplementationOnce(
        () =>
          '/var/lib/node/node_modules/@whook/whook/dist/handlers/getPing.js',
      );
      importer.mockResolvedValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [
          '/var/lib/node/node_modules/@whook/whook/dist',
          '/var/lib/node/node_modules/@whook/lol/dist',
        ],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('with no wrappers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      importer.mockImplementationOnce(async () => ({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      }));

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('getPing');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('with empty wrappers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        WRAPPERS: [],
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      importer.mockResolvedValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('getPingWrapped');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('for wrapped handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        WRAPPERS: [identity],
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockReturnValueOnce(
        '/home/whoami/my-whook-project/src/handlers/getPing.js',
      );
      importer.mockResolvedValueOnce({
        default: service(async () => async () => ({ status: 200 }), 'getPing'),
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });
      const result = await $autoload('getPingWrapped');

      expect({
        result,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
        injectorCalls: $injector.mock.calls,
        importerCalls: importer.mock.calls,
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });
  });

  describe('should fail', () => {
    it('with unexisting handlers', async () => {
      $injector.mockResolvedValueOnce({
        CONFIGS: {
          CONFIG: {
            testConfig: 'test',
          },
        },
      });
      $injector.mockResolvedValueOnce({
        API: { info: {} },
      });
      resolve.mockImplementationOnce(() => {
        throw new YError('E_NO_MODULE');
      });

      const $autoload = await initAutoload({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS_PATHS: [],
        $injector,
        SERVICE_NAME_MAP: {},
        INITIALIZER_PATH_MAP: {},
        WRAPPERS: [],
        log,
        importer,
        resolve: resolve as unknown as RequireResolve,
      });

      try {
        await $autoload('getPing');
        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: (err as YError).code,
          errorParams: (err as YError).params,
          logCalls: log.mock.calls.filter((args) => 'debug-stack' !== args[0]),
          injectorCalls: $injector.mock.calls,
          importerCalls: importer.mock.calls,
          resolveCalls: resolve.mock.calls,
        }).toMatchSnapshot();
      }
    });
  });
});
