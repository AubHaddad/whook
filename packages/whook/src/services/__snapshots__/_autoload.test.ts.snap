// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$autoload should fail with unexisting handlers 1`] = `
{
  "errorCode": "E_UNMATCHED_DEPENDENCY",
  "errorParams": [
    "getPing",
  ],
  "importerCalls": [],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸš« - Service "getPing" not found in "/home/whoami/my-whook-project/src/handlers/getPing".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
}
`;

exports[`$autoload should work for API 1`] = `
{
  "importerCalls": [
    [
      "/home/whoami/my-whook-project/src/services/API.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "API" found in "/home/whoami/my-whook-project/src/services/API.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "API" initializer from "/home/whoami/my-whook-project/src/services/API.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/services/API",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "API",
    "path": "/home/whoami/my-whook-project/src/services/API.js",
  },
}
`;

exports[`$autoload should work for SERVICE_NAME_MAP 1`] = `
{
  "importerCalls": [
    [
      "/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "SERVICE_NAME_MAP" found in "/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "SERVICE_NAME_MAP" initializer from "/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "SERVICE_NAME_MAP",
    "path": "/home/whoami/my-whook-project/src/services/SERVICE_NAME_MAP.js",
  },
}
`;

exports[`$autoload should work for a config constant 1`] = `
{
  "importerCalls": [],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
  ],
  "resolveCalls": [],
  "result": {
    "initializer": {
      "$name": "CONFIG",
      "$singleton": true,
      "$type": "constant",
      "$value": {
        "testConfig": "test",
      },
    },
    "name": "CONFIG",
    "path": "internal://CONFIG",
  },
}
`;

exports[`$autoload should work for configs 1`] = `
{
  "importerCalls": [],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
  ],
  "resolveCalls": [],
  "result": {
    "initializer": {
      "$name": "SERVICE_NAME_MAP",
      "$singleton": true,
      "$type": "constant",
      "$value": {},
    },
    "name": "SERVICE_NAME_MAP",
    "path": "internal://SERVICE_NAME_MAP",
  },
}
`;

exports[`$autoload should work for handlers 1`] = `
{
  "importerCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "getPing" found in "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "getPing" initializer from "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "getPing",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work for handlers hash 1`] = `
{
  "HANDLERS": {
    "getPing": [Function],
  },
  "importerCalls": [],
  "injectorCalls": [
    [
      [
        "API",
      ],
    ],
  ],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
  ],
  "resolveCalls": [],
  "result": {
    "initializer": [Function],
    "name": "HANDLERS",
    "path": "internal://HANDLERS",
  },
}
`;

exports[`$autoload should work for handlers in sub plugins 1`] = `
{
  "importerCalls": [
    [
      "/var/lib/node/node_modules/@whook/whook/dist/handlers/getPing.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸš« - Service "getPing" not found in "/home/whoami/my-whook-project/src/handlers/getPing".",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "getPing" found in "/var/lib/node/node_modules/@whook/whook/dist/handlers/getPing.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "getPing" initializer from "/var/lib/node/node_modules/@whook/whook/dist/handlers/getPing.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
    [
      "/var/lib/node/node_modules/@whook/whook/dist/handlers/getPing",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "getPing",
    "path": "/var/lib/node/node_modules/@whook/whook/dist/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work for name mapped handlers 1`] = `
{
  "importerCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPingMock.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸ“– - Using SERVICE_NAME_MAP to route "getPing" to "getPingMock".",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "getPingMock" found in "/home/whoami/my-whook-project/src/handlers/getPingMock.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "getPing" initializer via "getPingMock" resolution from "/home/whoami/my-whook-project/src/handlers/getPingMock.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPingMock",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "getPingMock",
    "path": "/home/whoami/my-whook-project/src/handlers/getPingMock.js",
  },
}
`;

exports[`$autoload should work for name mapped handlers 2`] = `
{
  "importerCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPingMock.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸ“– - Using SERVICE_NAME_MAP to route "getPing" to "getPingMock".",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "getPingMock" found in "/home/whoami/my-whook-project/src/handlers/getPingMock.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "getPing" initializer via "getPingMock" resolution from "/home/whoami/my-whook-project/src/handlers/getPingMock.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPingMock",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "getPingMock",
    "path": "/home/whoami/my-whook-project/src/handlers/getPingMock.js",
  },
}
`;

exports[`$autoload should work for path mapped handlers 1`] = `
{
  "importerCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸ“– - Using INITIALIZER_PATH_MAP to resolve the "getPing" module path.",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "getPing" found in "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "getPing" initializer from "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-other-project/src/handlers/getPing.js",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "getPing",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work for wrapped handlers 1`] = `
{
  "importerCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "getPingWrapped" found in "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "getPingWrapped" initializer from "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "getPingWrapped",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work with empty wrappers 1`] = `
{
  "importerCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "getPingWrapped" found in "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "getPingWrapped" initializer from "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "getPingWrapped",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing.js",
  },
}
`;

exports[`$autoload should work with no wrappers 1`] = `
{
  "importerCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing.js",
    ],
  ],
  "injectorCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ¤– - Initializing the \`$autoload\` service.",
    ],
    [
      "debug",
      "ðŸ’¿ - Service "getPing" found in "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
    [
      "debug",
      "ðŸ’¿ - Loading "getPing" initializer from "/home/whoami/my-whook-project/src/handlers/getPing.js".",
    ],
  ],
  "resolveCalls": [
    [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": {
    "initializer": [Function],
    "name": "getPing",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing.js",
  },
}
`;
